File: android/app/src/main/java/com/openland/spacex/transport/WebSocketTransport.kt
package com.openland.spacex.transport

import android.content.Context
import com.openland.spacex.utils.DispatchQueue
import org.json.JSONObject

class WebSocketTransport {

    private val url: String
    private val context: Context
    private val token: String?
    private var ws: WebSocketStableClient? = null
    private val queue = DispatchQueue()
    private var nextId = 1
    private var connected = false
    private var liveOperations = mutableMapOf<String, PendingOperation>()
    private val statusCallback: (connected: Boolean) -> Unit

    constructor(context: Context, url: String, token: String?, statusCallback: (connected: Boolean) -> Unit) {
        this.context = context
        this.statusCallback = statusCallback
        this.url = url
        this.token = token
        queue.async {
            this.connect()
        }
    }

    private fun connect() {

        this.ws = WebSocketStableClient(context, url, mapOf("x-openland-token" to this.token), queue, object : WebSocketStableClientCallback {
            override fun onMessage(message: JSONObject) {
                this@WebSocketTransport.handleMessage(message)
            }

            override fun onConnected() {
                if (!this@WebSocketTransport.connected) {
                    statusCallback(true)
                    this@WebSocketTransport.connected = true
                }
            }

            override fun onDisconnected() {
                if (this@WebSocketTransport.connected) {
                    statusCallback(false)
                    this@WebSocketTransport.connected = false
                }
                if (this@WebSocketTransport.ws != null) {
                    this@WebSocketTransport.ws!!.close()
                    this@WebSocketTransport.ws = null
                }
                connect()
            }
        })
        for (p in this.liveOperations) {
            postMessage("start", p.value.query, p.value.id)
        }
    }

    //
    // External Methods
    //

    fun operation(operation: JSONObject, callback: TransportOperationCallback): RunningOperation {
        val id = nextId++.toString()
        val op = PendingOperation(id, operation, callback)
        queue.async {
            liveOperations[id] = op
            postMessage("start", operation, id)
        }
        return op
    }

    //
    // All methods are called in Dispatch Queue
    //

    private fun handleMessage(response: JSONObject) {
        val type = response.getString("type")
        if (type == "data") {

            // Handle data and resolver level errors
            val id = response.getString("id")
            val payload = response.getJSONObject("payload")
            val errors = payload.optJSONObject("error")
            val data = payload.optJSONObject("data")
            if (this.liveOperations.containsKey(id)) {
                if (errors != null) {
                    this.liveOperations[id]!!.callback.onError(errors)
                } else {
                    this.liveOperations[id]!!.callback.onResult(data!!)
                }
            }
        } else if (type == "error") {

            // Handle low level error
            val id = response.getString("id")
            val payload = response.getJSONObject("payload")
            if (this.liveOperations.containsKey(id)) {
                this.liveOperations[id]!!.callback.onError(payload)
            }

        } else if (type == "complete") {
            val id = response.getString("id")
            if (this.liveOperations.containsKey(id)) {
                this.liveOperations.remove(id)!!.callback.onCompleted()
            }
        } else {
            // Ignore unknown message
        }
    }

    private fun postMessage(type: String, payload: JSONObject, id: String) {
        this.ws!!.postMessage(JSONObject(mapOf("id" to id, "type" to type, "payload" to payload)))
    }

    private inner class PendingOperation(val id: String, var query: JSONObject, val callback: TransportOperationCallback) : RunningOperation {
        override fun cancel() {
            queue.async {
                liveOperations.remove(id)
                if (connected) {
                    postMessage("stop", JSONObject(), id)
                }
            }
        }

        override fun lazyUpdate(operation: JSONObject) {
            queue.async {
                query = operation
            }
        }
    }
}

File: ios/WebSocketTransport.swift
//
//  WebSocketTransport.swift
//  openland
//
//  Created by Steve Kite on 5/13/19.
//  Copyright © 2019 Openland. All rights reserved.
//

import Foundation
import SwiftyJSON

enum TransportResult {
  case result(data: JSON)
  case error(error: JSON)
  case completed
}

protocol RunningOperation {
  func cancel()
  func updateVariables(variables: JSON)
}

fileprivate class PendingOperation: RunningOperation {
  let id: String
  let operation: OperationDefinition
  var variables: JSON
  weak var parent: WebSocketTransport?
  let callback: (TransportResult) -> Void
  
  init(parent: WebSocketTransport, id: String, operation: OperationDefinition, variables: JSON, callback: @escaping (TransportResult) -> Void) {
    self.parent = parent
    self.id = id
    self.operation = operation
    self.variables = variables
    self.callback = callback
  }
  
  func cancel() {
    if let p = self.parent {
      p.liveOperations.removeValue(forKey: self.id)
      p.connection?.post(message: JSON([
        "type": "stop",
        "id": self.id
        ]))
    }
  }
  
  func updateVariables(variables: JSON) {
    self.variables = variables
  }
}

class WebSocketTransport: WebSocketApolloTransportDelegate {
  private static let nextIdQueue = DispatchQueue(label: "ws-next-id")
  private static var nextId: Int = 1
  
  let url: String
  let params: [String: String?]
  fileprivate var connection: WebSocketApolloTransport? = nil
  fileprivate var liveOperations: [String: PendingOperation] = [:]
  fileprivate var connected = false
  fileprivate let queue = DispatchQueue(label: "transport")
  
  init(url: String, params: [String: String?]) {
    self.url = url
    self.params = params
  }
  
  func connect() {
    queue.async { [weak self] in
      if let s = self {
        s.doConnect()
      }
    }
  }
  
  func operation(operation: OperationDefinition, variables: JSON, handler: @escaping (TransportResult) -> Void) -> RunningOperation {
    var id: String = "0"
    WebSocketTransport.nextIdQueue.sync {
      id = "\(WebSocketTransport.nextId)"
      WebSocketTransport.nextId += 1
    }
    let q = JSON([
      "type": "start",
      "id": id,
      "payload": [
        "query": operation.body,
        "name": operation.name,
        "variables": variables
      ]])
    let pending = PendingOperation(parent: self, id: id, operation: operation, variables: variables, callback: handler)
    queue.sync {
      self.liveOperations[id] = pending
      self.connection?.post(message: q)
    }
    return pending
  }
  
  private func doConnect() {
    self.connection = WebSocketApolloTransport(url: url, params: params)
    self.connection!.delegate = self
    self.connection!.connect()
    for l in liveOperations {
      let q =  JSON([
        "type": "start",
        "id": l.value.id,
        "payload": [
          "query": l.value.operation.body,
          "name": l.value.operation.name,
          "variables": l.value.variables
        ]])
      self.connection!.post(message: q)
    }
  }
  
  private func onMessage(message: JSON) {
    let type = message["type"].stringValue
    if (type == "data") {
      let id = message["id"].stringValue
      let payload = message["payload"]
      let error = payload["error"]
      let data = payload["data"]
      if error.exists() {
        self.liveOperations[id]?.callback(TransportResult.error(error: error))
      } else {
        self.liveOperations[id]?.callback(TransportResult.result(data: data))
      }
    } else if type == "error" {
      let id = message["id"].stringValue
      let payload = message["payload"]
      self.liveOperations[id]?.callback(TransportResult.error(error: payload))
    } else if type == "complete" {
      let id = message["id"].stringValue
      self.liveOperations[id]?.callback(TransportResult.completed)
      self.liveOperations.removeValue(forKey: id)
    }
  }
  
  private func onConnected() {
    // TODO: Notify App
  }
  
  private func onDisconnected() {
    // TODO: Notify Apps
    
    // Reconnect
    if self.connection != nil {
      self.connection?.close()
      self.connection = nil
    }
    doConnect()
  }
  
  //
  // Transport delegate
  //
  
  func onMessage(src: WebSocketApolloTransport, message: JSON) {
    queue.async { [weak self] in
      if let s = self {
        if s.connection === src {
          s.onMessage(message: message)
        }
      }
    }
  }
  
  func onConnected(src: WebSocketApolloTransport) {
    queue.async { [weak self] in
      if let s = self {
        if s.connection === src {
          s.onConnected()
        }
      }
    }
  }
  
  func onDisconnected(src: WebSocketApolloTransport) {
    queue.async { [weak self] in
      if let s = self {
        if s.connection === src {
          s.onDisconnected()
        }
      }
    }
  }
  
  func close() {
    queue.async { [weak self] in
      if let s = self {
        s.connection?.close()
        s.connection = nil
      }
    }
  }
}
